sdk-java/src/main/java/io/littlehorse/sdk/worker/internal/LHTaskExecutor.java

```diff
@@ -0,0 +1,238 @@
+
+/**
+ * Manages a queue and executor for scheduling and executing tasks in memory.
+ * Future versions will share the same queue for many workers.
+ * It handles retries with a specified delay.
+ * Internally, a {@link Semaphore} is used to prevent overwhelming the in-memory queue.
+ * Clients can control the allowed size of this queue by specifying {@code LHW_NUM_WORKER_THREADS}.
+ * Clients can also specify the number of threads dedicated to executing tasks.
+ *
+ */
+@Slf4j
+public class LHTaskExecutor {
+
+    private final ScheduledExecutorService pool;
+    private final Semaphore semaphore;
+    private static final int TOTAL_RETRIES = 5;
+
+    public LHTaskExecutor(int concurrency) {
+        this.pool = Executors.newScheduledThreadPool(concurrency);
+        this.semaphore = new Semaphore(concurrency);
+        Thread closeExecutorHook = new Thread(() -> {
+            try {
+                this.close(4, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        });
+        Runtime.getRuntime().addShutdownHook(closeExecutorHook);
+    }
+
+    /**
+     * Schedule a new task to be executed
+     */
+    public void submitTaskForExecution(
+            ScheduledTask scheduledTask,
+            LittleHorseStub specificStub,
+            List<VariableMapping> mappings,
+            Object executable,
+            Method taskMethod) {
+        try {
+            this.semaphore.acquire();
+        } catch (InterruptedException exn) {
+            throw new RuntimeException(exn);
+        }
+        this.pool.submit(() -> {
+            this.doTask(scheduledTask, specificStub, mappings, executable, taskMethod);
+        });
+    }
+
+    /**
+     * Gracefully shutdown
+     */
+    public boolean close(int timeout, TimeUnit timeUnit) throws InterruptedException {
+        pool.shutdown();
+        return pool.awaitTermination(timeout, timeUnit);
+    }
+
+    private void doTask(
+            ScheduledTask scheduledTask,
+            LittleHorseGrpc.LittleHorseStub specificStub,
+            List<VariableMapping> mappings,
+            Object executable,
+            Method taskMethod) {
+        ReportTaskRun result = executeTask(
+                scheduledTask, LHLibUtil.fromProtoTs(scheduledTask.getCreatedAt()), mappings, executable, taskMethod);
+        this.semaphore.release();
+        String wfRunId = LHLibUtil.getWfRunId(scheduledTask.getSource()).getId();
+        try {
+            log.debug("Going to report task for wfRun {}", wfRunId);
+            specificStub.reportTask(result, new ReportTaskObserver(result, TOTAL_RETRIES, specificStub));
+            log.debug("Successfully contacted LHServer on reportTask for wfRun {}", wfRunId);
+        } catch (Exception exn) {
+            log.warn("Failed to report task for wfRun {}: {}", wfRunId, exn.getMessage());
+            scheduleRetry(result, specificStub, TOTAL_RETRIES);
```

Comment:
We should not schedule retries with the specific stub but rather with the `bootstrap` stub. The GRPC clients can do retries automatically themselves, but the problem is if we are doing a rolling bounce of the servers, the Specific Server will be down for about 45 seconds. That means the Specific Stub won't work for 45 seconds. We should use the Bootstrap Stub which balances across all servers.